### PR TITLE
Dashboard: Fix bug with NaN in alerting

### LINKED ISSUE #21953: NaN values break Grafana Alerting Rules
**What happened**: [grafana alerting aggregation](https://grafana.com/docs/grafana/latest/alerting/rules/#query-condition-example) function **min()** returns math.MaxFloat64 as min value for series which consist of points with NaN values

**What you expected to happen**: it should return 0

**How to reproduce it (as minimally and precisely as possible)**:

- with grafana UI
    - create a widget with a query where denominator can be equal to 0, example for Prometheus Datasource _something/rate(counter[5m])_
    - create a grafana alert and use min() aggregation function in Conditions
    - press "Test Rule" button
    - "Condition[0]: Query Result" should contain points where all values are "null" (they are "NaN"s actually)
    - "Condition[0]: Eval: true, Metric: {instance="localhost:9090"}, Value: 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.000"
    - please see [the dashboard ](https://gist.github.com/a-melnyk/5e48bc63c10164c3ea9d18f079cdfb56)

- please see [the test](https://github.com/a-melnyk/grafana/commit/360d92f5dc40bf549a372c4cc8ee3851922b2de2)

**Anything else we need to know?**: 
I see 2 ways to fix this:

- treat NaN values as invalid values (similarly to nil) in null.FloatFrom(), please see [the commit](https://github.com/a-melnyk/grafana/commit/918a9a6517ba22f39b41293cc16450a6da122acd)
- add an extra check in min() reducer, this seems to be hacky. Please see [the commit](https://github.com/a-melnyk/grafana/commit/ed61925a10163ef99589c3575d2fb69ca7e2fc20)

**Environment**:
- Grafana version: 6.5.2, 6.6.0, master
- Data source type & version: Prometheus 2.8.0, 2.15.2
- OS Grafana is installed on: official docker image
- Grafana plugins: none

### PR DESCRIPTION
**Pull request by a-melnyk**
**What this PR does / why we need it**: fixes issue with NaN float numbers in alerting aggregation functions (reducers)

**Which issue(s) this PR fixes**: 
Fixes #21953

### CODE REVIEW

File: pkg/services/alerting/conditions/reducer.go [hunk 1]
```diff
@@ -150,6 +150,13 @@ func calculateDiff(series *tsdb.TimeSeries, allNull bool, value float64, fn func
 	return allNull, value
 }
 
+func isValid(f null.Float) bool {
+	if f.Valid && !math.IsNaN(f.Float64) {
```
Comment (aknuds1) [inline pkg/services/alerting/conditions/reducer.go]:
```suggestion
	return f.Valid && !math.IsNaN(f.Float64)
```

File: pkg/services/alerting/conditions/reducer_test.go [hunk 2]
```diff
@@ -181,6 +182,21 @@ func TestSimpleReducer(t *testing.T) {
 
 			So(reducer.Reduce(series).Valid, ShouldEqual, false)
 		})
+
+		Convey("min should work with NaNs", func() {
+			result := testReducer("min", math.NaN(), math.NaN(), math.NaN())
+			So(result, ShouldEqual, float64(0))
+		})
+
+		Convey("isValid should treat NaN as not valid", func() {
```
Comment (aknuds1) [inline pkg/services/alerting/conditions/reducer_test.go]:
```suggestion
		Convey("isValid should treat NaN as invalid", func() {
```

File: pkg/services/alerting/conditions/reducer_test.go [hunk 3]
```diff
@@ -181,6 +182,21 @@ func TestSimpleReducer(t *testing.T) {
...
+			result := isValid(null.FloatFrom(math.NaN()))
+			So(result, ShouldEqual, false)
+		})
+
+		Convey("isValid should treat not vaild null.Float as not valid", func() {
```
Comment (aknuds1) [inline pkg/services/alerting/conditions/reducer_test.go]:
```suggestion
		Convey("isValid should treat invalid null.Float as invalid", func() {
```

File: pkg/services/alerting/conditions/reducer_test.go [hunk 4]
```diff
@@ -181,6 +182,21 @@ func TestSimpleReducer(t *testing.T) {
...
 		})
+
+		Convey("min should work with NaNs", func() {
+			result := testReducer("min", math.NaN(), math.NaN(), math.NaN())
+			So(result, ShouldEqual, float64(0))
+		})
+
+		Convey("isValid should treat NaN as not valid", func() {
+			result := isValid(null.FloatFrom(math.NaN()))
+			So(result, ShouldEqual, false)
```
Comment (aknuds1) [inline pkg/services/alerting/conditions/reducer_test.go]:
```suggestion
			So(result, ShouldBeFalse)
```

File: pkg/services/alerting/conditions/reducer_test.go [hunk 5]
```diff
@@ -181,6 +182,21 @@ func TestSimpleReducer(t *testing.T) {
...
+		})
+
+		Convey("isValid should treat not vaild null.Float as not valid", func() {
+			result := isValid(null.FloatFromPtr(nil))
+			So(result, ShouldEqual, false)
```
Comment (aknuds1) [inline pkg/services/alerting/conditions/reducer_test.go]:
```suggestion
			So(result, ShouldBeFalse)
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
I made some suggestions for improvement. @kylebrandt what do you think?

**Comment by kylebrandt** [comment]
> I made some suggestions for improvement. @kylebrandt what do you think?

They all look good to me (I don't have permission to commit the changes).

**Review by aknuds1 — APPROVED** [review]
LGTM!

**Pull request merged**